source/common/json/json_internal.cc

```diff
@@ -787,7 +791,11 @@ ObjectSharedPtr Factory::loadFromProtobufStruct(const ProtobufWkt::Struct& proto
 
 std::string Factory::serialize(absl::string_view str) {
   nlohmann::json j(str);
-  return j.dump();
+  // Call with defaults except in the case of UTF-8 errors which we may replace
+  // invalid UTF-8 characters instead of throwing an exception.
+  std::string ret = j.dump(-1, ' ', false, nlohmann::json::error_handler_t::replace);
+  ASSERT(ret.find(UnexpectedChar) == std::string::npos, "invalid UTF-8 is not allowed");
```

Comment:
Good call adding a comment to the assert. However, I would suggest this wording:
```
absl::StrCat("Invalid utf-8 `", str, "' leaked through utf8_range::IsStructurallyValid);
```
and we should aggressively fuzz this code in this PR to ensure this does not happen. When I was working on the sanitizer earlier -- especially my hand-rolled one that we wound up not using, I ran fuzzers for a day to see if they could catch any inconsistencies.